### ISSUE TITLE
[Models] Add decide primitive: typed decisions with calibrated probabilities

### ISSUE THREAD

**Issue by heskew** [HarperFast/harper#2779]
## Feature Summary

Add `decide` as a third model primitive alongside `embed` and `generate`: typed decisions with probabilities, as a first-class part of the models API (#510).

```ts
const d = await models.decide(ticket.body, {
	enum: ['billing', 'refund', 'bug', 'other'],
	description: 'Which queue should handle this support ticket?',
});
// { value: 'refund', probability: 0.91, distribution: [...], calibrated: true }
```

It would come with a new backend kind (`decision`), matching capability flags, a `@decide` schema directive (the write-time counterpart to `@embed`), and an optional calibration layer that learns from recorded outcomes.

## Problem This Solves

Apps constantly make fuzzy judgments: classify, route, score, moderate, triage, guardrail, and "is this a duplicate?" Today the only core option is `generate` with `responseFormat: { schema }`. That fits this kind of work poorly:

- **No probabilities.** The caller gets one answer with no confidence. So it can't threshold ("auto-route above 0.9, send the rest to a human"), and it can't find the low-confidence records later.
- **Wrong cost and latency for the job.** Generating tokens to pick one of four labels costs seconds and generation-level pricing. That rules out doing it on every write.
- **Free text to parse.** Even with a JSON schema, the backend returns text that has to be validated. Output that doesn't fit the schema becomes an error path.
- **Backends that can't generate have nowhere to go.** Several model families are built for exactly this and don't generate text:
  - fine-tuned classifiers
  - zero-shot NLI classifiers
  - cross-encoders
  - embedding-based classifiers
  - a new class of hosted "System One" decision models, e.g. TypeSafe's Jev, launched September 2026

  Registering any of them today means pretending to be a `generative` backend and claiming a `generate` capability they don't have. They also drop their most valuable output, the probabilities.

`embed` returns a representation (a vector) and `generate` returns open-ended text. Choosing from a closed, schema-defined set and returning a distribution over it is a different job, so it needs its own primitive.

### 1. `models.decide`

```ts
interface Models {
	// existing: embed, generate, generateStream, registerBackend, defineBackend, registerRouter
	decide<T = unknown>(state: DecideInput, schema: DecisionSchema, opts?: DecideOpts): Promise<Decision<T>>;
}

/** Text, or structured program state (JSON-serializable). */
type DecideInput = string | object;

/** Deliberately small JSON Schema subset, so that nearly every backend family can implement it. */
type DecisionSchema = DecisionLeaf | {
	type: 'object';
	description?: string;
	properties: Record<string, DecisionLeaf>; // one level only
};
type DecisionLeaf = { description?: string } & (
	| { enum: readonly (string | number | boolean)[] } // 2..255 values
	| { type: 'boolean' }
	| { type: 'integer'; minimum: number; maximum: number } // range of at most 255 values
);

type DecideOpts = {
	model?: string; // logical name, resolved via models.decision.<name>
	requires?: Capability[]; // e.g. ['calibrated']
	instructions?: string; // task framing beyond the schema's descriptions
	signal?: AbortSignal;
};

interface Decision<T> {
	id: string; // references the hdb_model_calls row; used for outcome recording
	value: T; // argmax (per-field argmax for object schemas)
	probability?: number; // P(value) for leaf schemas
	distribution?: Array<{ value: unknown; probability: number }>; // leaf schemas, sorted descending
	fields?: Record<string, Decision<unknown>>; // object schemas: per-field marginals
	calibrated: boolean; // whether the probabilities come from a calibrated source
	usage?: TokenUsage;
}
```

Why the design is shaped this way:

- **The schema is a required positional argument, not an option.** The schema is what defines a decision; if it were optional, the call would just be `generate`.
- **Always return the distribution.** Never return a bare value; the probabilities are the point.
- **Object schemas return per-field marginals, not joint probabilities.** Every backend family below can produce marginals. Only some can produce a joint distribution.
- **Small schema language on purpose.** Leaf types are enum, boolean and bounded integer, and objects nest only one level. Extracting spans of the input and nested arrays are out of scope for v1, because they would split backends into those that can and those that can't.

### 2. Backend kind, capabilities, config

This extends the existing registry (#1325), router (#1326) and module-backend loading (#1471). No new mechanisms are needed.

```ts
interface ModelBackend {
	// existing: embed?, generate?, generateStream?
	decide?(state: DecideInput, schema: DecisionSchema, opts: BackendOpts<DecideOpts>): Promise<ModelCallResult<Decision<unknown>>>;
}
interface ModelCapabilities {
	// existing: embed, generate, stream, tools, adapters
	decide: boolean;
	calibrated: boolean; // backend's probabilities are calibrated as returned
}
// registerBackend(kind: 'embedding' | 'generative' | 'decision', id, backend)
// defineBackend({ name, decide, calibrated? })
```

```yaml
models:
  decision:
    default:
      backend: '@acme/harper-decision-jev' # third-party module backend (#1471)
      apiKey: ${JEV_API_KEY}
      fallback: [llm]
    llm:
      backend: generative # built-in adapter over a generative logical name
      generative: default
```

Everything that follows from the existing models layer applies unchanged: routing and fallback, `requires` filtering, abort signals, tenant/app accounting, and `hdb_model_calls` rows (`method: 'decide'`) plus `model-decide` analytics.

### 3. Built-in adapters

These make `decide` useful without a dedicated decision backend. Each is opt-in through explicit config and never inferred.

- **`backend: generative`** implements `decide` on top of any configured generative backend:
  - By default it samples N times with the schema passed as `responseFormat` and takes a vote. It reports the vote frequencies as the distribution, with `calibrated: false`.
  - Where the provider exposes token log-probabilities, it scores each allowed value directly instead: one call instead of N, with better-quality probabilities. This needs an optional generative-backend hook, e.g. `scoreChoices?()`.
- **`backend: embedding`** (possible follow-up) implements `decide` as a nearest-centroid or logistic classifier over a configured embedding backend and a table of labeled examples. This is cheap, can run fully locally, and works well for stable label sets.

### 4. `@decide` schema directive

This is the write-time counterpart to `@embed` (#747), with the same trigger semantics: it runs before commit when the source field is present in the write, it's skipped on replication receive and log replay, and a source set to `null` clears the value.

```graphql
type Ticket @table @export {
	id: ID @primaryKey
	body: String
	route: String @decide(source: "body", values: ["billing", "refund", "bug", "other"], model: "default", confidence: "routeConfidence")
	routeConfidence: Float @indexed
}
```

Storing the confidence in an indexed attribute makes confidence queryable. A review queue becomes an ordinary search (`routeConfidence < 0.7`), and changing the threshold doesn't require re-running the model.

### 5. Calibration from outcomes (later phase)

Most backends' raw scores are not calibrated. A backend-agnostic layer fixes that over time, using data Harper already holds:

- `models.recordOutcome(decision.id, actualValue)` records the real answer, e.g. from a human review or a downstream result.
- A periodic job fits a correction per (backend, schema, field): temperature scaling for multi-class schemas, Platt or isotonic scaling for binary ones. The correction is applied on return. Once a configurable sample threshold is met, results report `calibrated: true`.
- It also exposes reliability stats (expected calibration error, sample counts) in analytics, so operators can see how trustworthy the probabilities are.

This is the strongest reason for the primitive to live in the database rather than a vendor SDK: the database is where the outcomes end up.

### Suggested phasing

1. Types, facade `decide`, `decision` registry kind, capabilities, `defineBackend`, config, analytics, and a TestBackend.
2. The `generative` adapter (vote first, then log-probability scoring). Vendor decision backends ship as module backends via #1471.
3. The `@decide` directive.
4. Outcome recording and calibration.

## User Stories

- As an app developer, I want to route every incoming support ticket at write time and send only low-confidence ones to a human, so that triage is automatic without being blind.
- As a developer building on the agent loop (#612), I want a cheap calibrated check before an expensive `generate` call (is the request in scope? which tool family?), so that most traffic never reaches the LLM.
- As a developer, I want to guardrail or score LLM output (policy check, groundedness, relevance) with a probability, so that I can choose the threshold instead of trusting a yes/no.
- As an operator, I want to swap a hosted decision model for a local classifier, or the reverse, by changing config, so that the app code and schema never change. This is the same source-portability argument #510 makes for `embed` and `generate`.
- As an operator, I want to know whether the confidence numbers mean anything, so that thresholds are set against measured calibration rather than guesses.

## Alternatives Considered

- **`generate` + `responseFormat: { schema }`.** It works today, but it has no probabilities, costs a full generation, can fail to match the schema, and excludes backends that can't generate text.
- **Adding `logprobs` to `GenerateResult`.** This helps LLM backends only. It ties the feature to text generation and leaves classifiers, NLI, cross-encoders and decision models out. (A log-probability hook is still useful inside the `generative` adapter.)
- **A plugin that exposes its own global.** This fragments the API and loses routing, fallback, accounting and a directive to fill fields on write. It's the same fork #510 was written to prevent.
- **Other names.**
  - `classify` is too narrow: it doesn't cover scoring or bounded-integer outputs.
  - `predict` is vague, and people read it as forecasting.
  - `judge` only covers guardrails.
  - `infer` describes every model call.
  - `decide` matches how both application code and this class of model describe the operation.

## Priority/Impact

Medium - important but not urgent

## Examples or References

- [TypeSafe AI: Introducing System One models and Jev](https://typesafe.ai/blog/introducing-system-one-models-and-jev): a hosted model class built for this operation (typed outputs with calibrated probabilities)
- Existing backend families that can implement `decide`:
  - zero-shot NLI classification (e.g. `facebook/bart-large-mnli`)
  - fine-tuned small classifiers (DeBERTa, SetFit)
  - cross-encoders
  - grammar-restricted LLM decoding (llama.cpp grammars, vLLM guided decoding)
- Guo et al., [On Calibration of Modern Neural Networks](https://arxiv.org/abs/1706.04599) (2017): temperature scaling
- Related: #510, #747 (`@embed`), #1325, #1326, #1471, #612, #750, #1684

## Additional Context

Open questions:

- **Schema language.** A JSON Schema subset (as proposed) or Harper's own type system? And should `@decide` read allowed values from a GraphQL enum type instead of a `values:` argument?
- **Object schemas.** Are per-field marginals enough for v1, or is there a real need for joint probabilities?
- **Where outcomes live.** Columns on `hdb_model_calls`, or a separate table? Retention needs differ: calibration wants a longer history than per-call forensics.
- **Write-path latency.** Should `@decide` allow the vote-based `generative` adapter (N calls on every write), or require a backend with the `decide` capability built in? The derived caching table in #750 may apply here too.
- **Cardinality cap.** Should the 255-value cap be fixed, or depend on the backend?
- **`/v1` gateway.** OpenAI's API has no equivalent endpoint, so the proposal is to leave the gateway unchanged for now.

## Are you planning to fix this issue?

Maybe, depending on complexity

---
🤖 Filed by Claude Code (Opus 5.5) on behalf of @heskew

**Comment by IRONICBo** [comment]
One contract gap: the schema can force a closed-world answer, but low probability is not the same as “none of these.” Because the distribution normalizes across the enum, an out-of-set input can still produce a confident argmax.

I would make `no_match` / abstention first-class in `DecisionSchema` and `Decision`, let `recordOutcome` accept an out-of-set result, and report selective risk versus coverage at each automation threshold rather than ECE alone. In Jev Social we keep `unsupported` as an explicit route and remove already-attempted actions; that boundary prevents a calibrated-looking distribution from authorizing a nonsensical operation. A backend that cannot represent abstention should advertise that limitation instead of silently forcing a choice.
